### PR TITLE
Add microbenchmark for core UnmanagedMemoryStream scenario

### PR DESCRIPTION
**Pull request by jkotas**
Related to https://github.com/dotnet/runtime/pull/93766

### CODE REVIEW

File: src/benchmarks/micro/libraries/System.IO/UnmanagedMemoryStreamTests.cs [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+        [GlobalCleanup]
+        public void Cleanup()
+        {
+            Marshal.FreeCoTaskMem((IntPtr)buffer);
+        }
+
+        [Benchmark]
+        public void ReadAsBytes()
+        {
+            using (var ums = new UnmanagedMemoryStream(buffer, Length))
```
Comment (jkotas) [inline src/benchmarks/micro/libraries/System.IO/UnmanagedMemoryStreamTests.cs:33]: I am not sure whether this should better be in IterationSetup. The docs for [IterationSetup](https://github.com/dotnet/BenchmarkDotNet/blob/e170684208103ca5ba4212ad8dc7c2aad5cf02d4/src/BenchmarkDotNet.Annotations/Attributes/IterationSetupAttribute.cs#L7) say that it should not be used for microbenchmarks.

Comment (cincuranet) [reply src/benchmarks/micro/libraries/System.IO/UnmanagedMemoryStreamTests.cs:33]: As long as the benchmark takes roughly more than 100ms the IterationSetup is fine.

Comment (jkotas) [reply src/benchmarks/micro/libraries/System.IO/UnmanagedMemoryStreamTests.cs:33]: This microbenchmark takes a lot less than 100ms and I do not think it makes sense to artificially inflate it to take more than 100ms.

File: src/benchmarks/micro/libraries/System.IO/UnmanagedMemoryStreamTests.cs [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    [BenchmarkCategory(Categories.Libraries)]
+    public unsafe class UnmanagedMemoryStreamTests
+    {
+        private const int Length = 10000;
+        private byte* buffer;
+
+        [GlobalSetup]
+        public void Setup()
+        {
+            buffer = (byte*)Marshal.AllocCoTaskMem(Length);
```
Comment (Jozkee) [inline src/benchmarks/micro/libraries/System.IO/UnmanagedMemoryStreamTests.cs:21]: Just wondering why you chose this over AllocHGlobal?

**Review by Jozkee — APPROVED** [review]

File: src/benchmarks/micro/libraries/System.IO/UnmanagedMemoryStreamTests.cs [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+using BenchmarkDotNet.Attributes;
+using MicroBenchmarks;
+
+namespace System.IO.Tests
+{
+    [BenchmarkCategory(Categories.Libraries)]
+    public unsafe class UnmanagedMemoryStreamTests
+    {
+        private const int Length = 10000;
+        private byte* buffer;
```
Comment (Jozkee) [inline src/benchmarks/micro/libraries/System.IO/UnmanagedMemoryStreamTests.cs:16]:
```suggestion
        private byte* _buffer;
```

File: src/benchmarks/micro/libraries/System.IO/UnmanagedMemoryStreamTests.cs (hunk 2 above)
Comment (jkotas) [reply src/benchmarks/micro/libraries/System.IO/UnmanagedMemoryStreamTests.cs]: No particular reason. Either of these two would work just fine for this test.

File: src/benchmarks/micro/libraries/System.IO/UnmanagedMemoryStreamTests.cs [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+                while (ums.ReadByte() >= 0)
+                {
+                }
+            }
+        }
+
+        [Benchmark]
+        public void ReadAsArrays()
+        {
+            var array = new byte[1];
```
Comment (jkotas) [inline src/benchmarks/micro/libraries/System.IO/UnmanagedMemoryStreamTests.cs:44]:
```suggestion
            var array = new byte[128];
```

**Review by cincuranet — APPROVED** [review]

File: src/benchmarks/micro/libraries/System.IO/UnmanagedMemoryStreamTests.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (jkotas) [reply src/benchmarks/micro/libraries/System.IO/UnmanagedMemoryStreamTests.cs]: This test is meant to test raw cost of UnmanagedMemoryStream. With larger array sizes, this test would be testing performance of memory copy and the raw cost of UnmanagedMemoryStream would be diminished.

I have changed it to 128. It is still small enough to make the raw cost of UnmanagedMemoryStream dominate.

[the remark this answers is not in the document]
Comment (cincuranet) [reply src/benchmarks/micro/libraries/System.IO/UnmanagedMemoryStreamTests.cs]: Makes sense. Thanks.

**Comment by cincuranet** [comment]
Failures are unrelated.

**Pull request merged**